cranelift/codegen/src/isa/aarch64/abi.rs

```diff
@@ -1323,286 +607,138 @@ impl ABIBody for AArch64ABIBody {
+            r.get_hw_encoding() >= 8 && r.get_hw_encoding() <= 15
+        }
+        _ => panic!("Unexpected RegClass"),
     }
+}
 
-    fn emit_copy_reg_to_arg<C: LowerCtx<I = Self::I>>(
-        &self,
-        ctx: &mut C,
-        idx: usize,
-        from_reg: Reg,
-    ) {
-        match &self.sig.args[idx] {
-            &ABIArg::Reg(reg, ty, ext)
-                if ext != ir::ArgumentExtension::None && ty_bits(ty) < 64 =>
-            {
-                assert_eq!(RegClass::I64, reg.get_class());
-                let signed = match ext {
-                    ir::ArgumentExtension::Uext => false,
-                    ir::ArgumentExtension::Sext => true,
-                    _ => unreachable!(),
-                };
-                ctx.emit(Inst::Extend {
-                    rd: Writable::from_reg(reg.to_reg()),
-                    rn: from_reg,
-                    signed,
-                    from_bits: ty_bits(ty) as u8,
-                    to_bits: 64,
-                });
-            }
-            &ABIArg::Reg(reg, ty, _) => {
-                ctx.emit(Inst::gen_move(
-                    Writable::from_reg(reg.to_reg()),
-                    from_reg,
-                    ty,
-                ));
-            }
-            &ABIArg::Stack(off, ty, ext) => {
-                if ext != ir::ArgumentExtension::None && ty_bits(ty) < 64 {
-                    assert_eq!(RegClass::I64, from_reg.get_class());
-                    let signed = match ext {
-                        ir::ArgumentExtension::Uext => false,
-                        ir::ArgumentExtension::Sext => true,
-                        _ => unreachable!(),
-                    };
-                    // Extend in place in the source register. Our convention is to
-                    // treat high bits as undefined for values in registers, so this
-                    // is safe, even for an argument that is nominally read-only.
-                    ctx.emit(Inst::Extend {
-                        rd: Writable::from_reg(from_reg),
-                        rn: from_reg,
-                        signed,
-                        from_bits: ty_bits(ty) as u8,
-                        to_bits: 64,
-                    });
-                }
-                ctx.emit(store_stack(MemArg::SPOffset(off, ty), from_reg, ty))
+fn get_callee_saves(
+    call_conv: isa::CallConv,
+    regs: &Set<Writable<RealReg>>,
+) -> (Vec<Writable<RealReg>>, Vec<Writable<RealReg>>) {
+    let mut int_saves = vec![];
+    let mut vec_saves = vec![];
+    for &reg in regs.iter() {
+        if is_callee_save(call_conv, reg.to_reg()) {
+            match reg.to_reg().get_class() {
+                RegClass::I64 => int_saves.push(reg),
+                RegClass::V128 => vec_saves.push(reg),
+                _ => panic!("Unexpected RegClass"),
             }
         }
     }
+    // Sort registers for deterministic code output.
+    int_saves.sort_by_key(|r| r.to_reg().get_index());
```

Comment:
Can these be unstable sorts, in the interest of cranking every last nanosecond out of the machine?

Reply:
I thought about that, but IMHO it's better to have truly deterministic output -- not doing so could create all sorts of weird testing issues, problems with caching at various levels, etc. We expect these lists to be short (at most the number of machine registers) so the difference shouldn't be too large...

Reply:
(And to be clear, the nondeterminism is otherwise new with this patch, because it changes the iteration to be directly over the `Set` rather than over the set-converted-to-vec, and the iteration order is unspecified.)

Reply:
Oh, yes, I am all in favour of determinism.  I retract my comment :-)

Reply:
`unstable_sort_by_key` doesn't affect determinism. It only causes a difference when multiple values have the same sort key. In that case the order would be dependent on the internals of the sorting algorithm, but not any random source. `sort_by_key` on the other hand guarantees that multiple values with the same sort key are kept in the same order.